catalyst-gateway/bin/src/db/event/signed_docs/mod.rs

```diff
@@ -20,13 +59,137 @@ const INSERT_SIGNED_DOCS: &str = include_str!("./sql/insert_signed_documents.sql
 ///  - `ver` is a UUID v7
 ///  - `doc_type` is a UUID v4
 #[allow(dead_code)]
-pub(crate) async fn insert_signed_docs(
-    id: &uuid::Uuid, ver: &uuid::Uuid, doc_type: &uuid::Uuid, author: &String,
-    metadata: &Option<serde_json::Value>, payload: &Option<serde_json::Value>, raw: &Vec<u8>,
-) -> anyhow::Result<()> {
+pub(crate) async fn insert_signed_docs(doc: &FullSignedDoc) -> anyhow::Result<()> {
+    match select_signed_docs(&doc.body.id, &Some(doc.body.ver)).await {
+        Ok(res_doc) => {
+            anyhow::ensure!(
+                &res_doc == doc,
+                "Document with the same `id` and `ver` already exists"
+            );
+            return Ok(());
+        },
+        Err(err) if err.is::<NotFoundError>() => {},
+        Err(err) => return Err(err),
+    }
+
     EventDB::modify(INSERT_SIGNED_DOCS, &[
-        id, ver, doc_type, author, metadata, payload, raw,
+        &doc.body.id,
+        &doc.body.ver,
+        &doc.body.doc_type,
+        &doc.body.author,
+        &doc.body.metadata,
+        &doc.payload,
+        &doc.raw,
     ])
     .await?;
     Ok(())
 }
+
+/// Make a select query into the `event-db` by getting data from the `signed_docs` table.
+///
+/// * This returns a single document. All data from the document is returned, including
+///   the `payload` and `raw` fields.
+/// * `ver` should be able to be optional, in which case get the latest ver of the given
+///   `id`.
+///
+/// # Arguments:
+///  - `id` is a UUID v7
+///  - `ver` is a UUID v7
+pub(crate) async fn select_signed_docs(
+    id: &uuid::Uuid, ver: &Option<uuid::Uuid>,
+) -> anyhow::Result<FullSignedDoc> {
+    let query_template = get_template(&SELECT_SIGNED_DOCS_TEMPLATE)?;
+    let query = query_template.render(serde_json::json!({
+        "id": id,
+        "ver": ver,
+    }))?;
+    let res = EventDB::query_one(&query, &[]).await?;
+
+    let ver = if let Some(ver) = ver {
+        *ver
+    } else {
+        res.try_get("ver")?
+    };
+
+    let doc_type = res.try_get("type")?;
+    let author = res.try_get("author")?;
+    let metadata = res.try_get("metadata")?;
+    let payload = res.try_get("payload")?;
+    let raw = res.try_get("raw")?;
+
+    Ok(FullSignedDoc {
```

Comment:
Don't create types like this inline, put the above into a function implemented for the `FullSignedDoc` type which implements this logic and returns the type, or an error.

Maybe something like:
```
  SignedDocBody::from_query_result(res,ver)
```
or
```
  SignedDocBody::new(res,ver)
```
But if we use `new` consider if this type is going to be `newed` by some other code, and not from db result, in which case, which one should be `new` and which one needs a custom named function to create the type?